src/rgw/rgw_dedup_store.cc

```diff
@@ -0,0 +1,847 @@
+    }
+    else {
+      if (ret == 0) {
+        // no error reported, but we wrote nothing which should never happen
+        ldpp_dout(dpp, 5) << __func__ << "::ERR: No Data was written to " << oid
+                          << ", bl.length()=" << bl.length() << dendl;
+        ret = -ENODATA;
+      }
+      ldpp_dout(dpp, 1) << "ERROR: failed to write " << oid
+                        << " with: " << cpp_strerror(ret) << dendl;
+    }
+
+    return ret;
+  }
+
+  //---------------------------------------------------------------------------
+  int disk_block_seq_t::flush(librados::IoCtx &ioctx)
+  {
+    unsigned len = (p_curr_block + 1 - p_arr) * sizeof(disk_block_t);
+    bufferlist bl = bufferlist::static_from_mem((char*)p_arr, len);
+    int ret = store_slab(ioctx, bl, d_md5_shard, d_worker_id, d_seq_number, dpp);
+    // TBD: Can we recycle the buffers?
+    // Need to make sure the call to rgw_put_system_obj was fully synchronous
+
+    // d_seq_number++ must be called **after** flush!!
+    d_seq_number++;
+    p_stats->egress_slabs++;
+    slab_reset();
+    return ret;
+  }
+
+  //---------------------------------------------------------------------------
+  static void __attribute__ ((noinline))
+  report_bad_access(const DoutPrefixProvider* dpp, const char *caller)
+  {
+    ldpp_dout(dpp, 0) << "::ERR: " << caller
+                      << "() was called on uninitialized disk_block_seq_t\n"
+                      << dendl;
+  }
+
+  //---------------------------------------------------------------------------
+  int disk_block_seq_t::flush_disk_records(librados::IoCtx &ioctx)
+  {
+    if (unlikely(!p_arr)) {
+      report_bad_access(dpp, __func__);
+      return -1;
+    }
+
+    ldpp_dout(dpp, 20) << __func__ << "::worker_id=" << (uint32_t)d_worker_id
+                       << ", md5_shard=" << (uint32_t)d_md5_shard << dendl;
+
+    // we need to force flush at the end of a cycle even if there was no work done
+    // it is used as a signal to worker in the next step
+    if (p_curr_block == &p_arr[0] && p_curr_block->is_empty()) {
+      ldpp_dout(dpp, 20) << __func__ << "::Empty buffers, generate terminating block" << dendl;
+    }
+    p_stats->egress_blocks++;
+    p_curr_block->close_block(dpp, false);
+
+    int ret = flush(ioctx);
+    return ret;
+  }
+
+  //---------------------------------------------------------------------------
+  int disk_block_seq_t::add_record(librados::IoCtx        &ioctx,
+                                   const rgw::sal::Bucket *p_bucket,
+                                   const rgw::sal::Object *p_obj,
+                                   const parsed_etag_t    *p_parsed_etag,
+                                   const std::string      &obj_name,
+                                   uint64_t                obj_size,
+                                   record_info_t          *p_rec_info) // OUT-PARAM
+  {
+    if (unlikely(!p_arr)) {
+      report_bad_access(dpp, __func__);
+      return -1;
```

Comment:
integer error codes that trickle in the system should have be standard linux values, `-1` does not make sense here